File: misc/bing_ip.py
#!/usr/bin/env python
#Use BING to search for websites hosted on a ip address and return the current ip they have
#Author: DiabloHorn (http://diablohorn.wordpress.com)

#Working with the bing engine
#example url: http://api.bing.net/xml.aspx?AppId=<APPID>&Version=2.2&Query=ip:74.207.254.18&Sources=web&web.count=50&web.offset=0
#example usage: nmap -vv -n -sL 74.207.254.15-20 | grep -o '[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}' | ./bing_ip.py -a <APPID> -f

import sys
import getopt
import urllib2
from xml.dom.minidom import parse, parseString
from urlparse import urlparse
import socket

class bing():
    _appid = ""
    _url = "http://api.bing.net/xml.aspx?"
    _urlparams = {"appid":"","version":"2.2","query":"","sources":"web","web.count":"50","web.offset":"0"}
    _foundurls = {}
    _totalres = 0
    
    def _constructurl(self):
        queryurl = self._url
        for k,v in self._urlparams.iteritems():
            if queryurl[-1] == "?":
                queryurl = queryurl + k + "=" + v
            else:
                queryurl = queryurl + "&" + k + "=" + v
        return queryurl
    
    def __init__(self,appid):
        self._urlparams.update({"appid":appid})
        
    def _fetchresult(self,bingurl):
        con = urllib2.urlopen(bingurl)
        content = con.read()
        con.close()
        return content
        
    def _query(self,qry,offset=0):
        self._urlparams.update({"query":qry,"web.offset":str(offset)})
        return self._fetchresult(self._constructurl())
    
    def _getUrls(self,bingdata,uptotal=False):
        dom = parseString(bingdata)
        if(uptotal):
            self._totalres = dom.getElementsByTagName("web:Total")[0].firstChild.data.encode('utf8')    
        weburls = dom.getElementsByTagName("web:Url")
        for url in weburls:
            host = urlparse(url.firstChild.data.encode('utf8')).netloc
            try:
                self._foundurls[host] = socket.gethostbyname(host)
            except:
                self._foundurls[host] = ""
                        
    def IPQuery(self,qry):
        self._foundurls.clear()
        self._getUrls(self._query("ip:"+qry),True)
        print "[*] Found Results(" + qry + "): " + self._totalres
        pages = int(self._totalres) / 50
        print "[*] Pages: " + str(pages)
        for i in range (0,pages):
            self._getUrls(self._query("ip:"+qry,(i+1)))
        return self._foundurls

def txthelp():
    print "[*] DiabloHorn http://diablohorn.wordpress.com"
    print "[*] Use BING to search for websites hosted on a ip address"
    print "[*] " + sys.argv[0] + " -a <appid> -i <ip> [-f <ipfile list>]"
    print "[*] -a BING appid"
    print "[*] -i ip address to use"
    print "[*] -f read file from stdin(ex: cat ips.txt | " + sys.argv[0] 
    print "[*] -h this help" 


    
#main :)
if __name__ == "__main__":
    if len(sys.argv) <=2:
        txthelp()
        sys.exit(0)
    
    targetscan = ""
    fromfile = False
    bingkey = ""
        
    try:
        opts, args = getopt.getopt(sys.argv[1:],"a:i:fh",["appid","ip","ipfile","help"])
    except getopt.GetoptError, err:
        print str(err)
        sys.exit(0)

    for o,a in opts:
        if o == "-h":
            txthelp()
            sys.exit(0)            
        elif o == "-v": 
            verbose = True
        elif o == "-a":
            bingkey = a            
        elif o == "-i":
            targetscan = a
        elif o == "-f":
            fromfile = True
        else:
            txthelp()
            sys.exit(0)

    bip = bing(bingkey)
    if fromfile:
        for line in sys.stdin:
            for k,v in bip.IPQuery(line.strip()).iteritems():
                print k+":"+v
    else:
        for k,v in bip.IPQuery(targetscan).iteritems():
            print k+":"+v

File: misc/bing_ip.rb
#!/usr/bin/env ruby
#Use BING to search for websites hosted on a ip address and return the current ip they have
#Author: DiabloHorn (http://diablohorn.wordpress.com)
#does the same as the python version, was curious about ruby cli coding

#Working with the bing engine
#example url: http://api.bing.net/xml.aspx?AppId=<APPID>&Version=2.2&Query=ip:74.207.254.18&Sources=web&web.count=50&web.offset=0

#even though ruby doesn't really have a main, according to wikipedia
# http://en.wikipedia.org/wiki/Main_function#Ruby

require 'optparse'
require 'net/http'
require 'uri'
require 'cgi'
require 'rexml/document'
require 'socket'


class Bing
    BURL = "http://api.bing.net"
    BPATH = "/xml.aspx"
    
    def initialize(bingid,bingurl=BURL)
        @params = {"appid" => "#{bingid}","Version" => "2.2","Query" => "","Sources" => "Web","web.count" => "50","web.offset" => "0"}
        @uri = URI.parse(bingurl)
        @foundurls = {}
    end
    
    private
    def page_content()
        page_url = BPATH + "?" + @params.collect{|k,v| "#{k}=#{CGI::escape(v.to_s)}"}.join('&')
        req = Net::HTTP::Get.new(page_url)
        http = Net::HTTP.new(@uri.host, @uri.port)
        response = http.start do |http| http.request(req) end
        response.body
    end
    
    private 
    def content_parse(xml,ttr=false)
    #http://www.componentworkshop.com/blog/2009/07/13/using-the-bing-xml-search-api-with-ruby
        doc = REXML::Document.new(xml)
        if(ttr)
            doc.elements.each("SearchResponse/web:Web") do |t| @totalRes = t.get_elements("web:Total")[0].text end
        end
        doc.elements.each("SearchResponse/web:Web/web:Results/web:WebResult") do |result|
            ruri = URI.parse(result.get_elements("web:Url")[0].text)
            begin
                @foundurls[ruri.host.to_s] = Socket.getaddrinfo(ruri.host.to_s,"http","AF_INET")[0][3]
            rescue
                @foundurls[ruri.host.to_s] = nil
            end
        end
    end
    
    public
    def ip_query(ip)
        @foundurls.clear
        @params["Query"] = "ip:"+ip
        @params["web.offset"] = "0"
        puts "Searching: #{@params["Query"]}"
        xmldata = page_content()
        content_parse(xmldata,true)
        pages = (@totalRes.to_i / 50)
        puts "Results Found: " + @totalRes
        puts "Pages: " + pages.to_s
        for i in (0..pages)
            @params["web.offset"] = (i+1).to_s
            content_parse(page_content())
        end
        @foundurls
    end
end

if __FILE__ == $PROGRAM_NAME
    #parse command line arguments
    options = {}
    optparse = OptionParser.new do |opts|
        opts.banner = "[*] DiabloHorn http://diablohorn.wordpress.com\n" + 
                        "[*] Use BING to search for websites hosted on a ip address\n" + 
                        "[*] " + $PROGRAM_NAME + " -a <appid> -i ([file:location] | [ip])\n"
    
        #application specific options
                #switch has mandatory argument 
        opts.on("-a", "--bingappid BINGAPPID", "BING AppID") do |a|
            options[:bappid] = a
        end  
                #switch has mandatory argument (file:[location], inspired by metasploit)
        opts.on("-i", "--ipaddress IP", "IP address (use file:[location] to use a file") do |i|
            options[:ip] = i
        end

        opts.on_tail("-v", "--[no-]verbose", "Run verbosely") do |v|
            options[:verbose] = v
        end
        
        opts.on_tail("-h", "--help", "Show this message") do
            puts opts
            exit
        end    
    end
    
    #http://stackoverflow.com/questions/1541294/how-do-you-specify-a-required-switch-not-argument-with-ruby-optionparser
    begin
        optparse.parse!
        mandatory = [:bappid,:ip]
        missing = mandatory.select{ |param| options[param].nil? }
        if not missing.empty?
            puts "Missing options: #{missing.join(', ')}"
            puts optparse
            exit                                                                                                                                     
        end
    rescue OptionParser::InvalidOption, OptionParser::MissingArgument
        puts $!.to_s
        puts optparse       
        exit
    end
    
    #start processing
    #first let's check if it's a file or a single ip.
    bs = Bing.new(options[:bappid])
    
    if options[:ip][0,5] == "file:"
        #get file path and cycle through it
        File.open(options[:ip][5,options[:ip].length], "r") do |infile|
            while (line = infile.gets)
                urlsfound = bs.ip_query(line.rstrip) 
                urlsfound.each do|k,v|
                    puts "#{k}:#{v}"
                end
            end
        end
    else
        #quick and painless one lookup
        urlsfound = bs.ip_query(options[:ip]) 
        urlsfound.each do|k,v|
            puts "#{k}:#{v}"
        end
    end
end
